From 89043be4ece938558b973e0bdf8c8de8ee160893 Mon Sep 17 00:00:00 2001
From: Shahed Shaikh <shahedx.shaikh@intel.com>
Date: Sun, 28 Jun 2020 23:23:49 +0800
Subject: [PATCH] DRV_DPM_SW-302: DPM optimize dp_notifier_invoke

In dp_notifier_invoke - in case of subif register/deregister
event notification, there is no need to call dp_get_netif_subifid_priv()
to retrieve subif_id. Simply pass subif_id while calling dp_notifier_invoke.

Signed-off-by: Shahed Shaikh <shahedx.shaikh@intel.com>
---
 drivers/net/datapath/dpm/datapath.h          |  2 +-
 drivers/net/datapath/dpm/datapath_api.c      | 16 ++++++++--------
 drivers/net/datapath/dpm/datapath_misc.c     |  4 ++--
 drivers/net/datapath/dpm/datapath_notifier.c | 13 ++-----------
 4 files changed, 13 insertions(+), 22 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index 534e2663b671..6f64b83a0407 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -1071,7 +1071,7 @@ static inline bool is_dsl(struct pmac_port_info *port)
 }
 
 int dp_notifier_invoke(int inst,
-		       struct net_device *dev, u32 port_id,
+		       struct net_device *dev, u32 port_id, u32 subif_id,
 		       enum DP_EVENT_TYPE type);
 
 int dp_dealloc_cqm_port(struct module *owner, u32 dev_port,
diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index f387ea47cba5..990fa47a29eb 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -435,7 +435,7 @@ static int32_t dp_alloc_port_private(int inst,
 			return DP_FAILURE;
 		}
 
-		dp_notifier_invoke(inst, dev, port_id,
+		dp_notifier_invoke(inst, dev, port_id, 0,
 				   DP_EVENT_DE_ALLOC_PORT);
 
 		cbm_data.deq_port = port->deq_port_base;
@@ -561,7 +561,7 @@ static int32_t dp_alloc_port_private(int inst,
 		 "Port %d allocation succeed for module %s with dev_port %d\n",
 		 port_id, owner->name, dev_port);
 
-	dp_notifier_invoke(inst, dev, port_id, DP_EVENT_ALLOC_PORT);
+	dp_notifier_invoke(inst, dev, port_id, 0, DP_EVENT_ALLOC_PORT);
 
 	return port_id;
 }
@@ -860,7 +860,7 @@ int32_t dp_deregister_subif_private(int inst, struct module *owner,
 	bp = sif->bp;
 
 	if (sif->ctp_dev)
-		dp_notifier_invoke(inst, sif->ctp_dev, port_id,
+		dp_notifier_invoke(inst, sif->ctp_dev, port_id, subif_id->subif,
 				   DP_EVENT_DE_REGISTER_SUBIF);
 
 	/* subif_hw_reset */
@@ -881,7 +881,7 @@ int32_t dp_deregister_subif_private(int inst, struct module *owner,
 
 		dp_inst_del_dev(dev, inst, port_id, subif_id->subif, 0);
 
-		dp_notifier_invoke(inst, dev, port_id,
+		dp_notifier_invoke(inst, dev, port_id, subif_id->subif,
 				   DP_EVENT_DE_REGISTER_SUBIF);
 	}
 	/* reset mib, flag, and others */
@@ -1105,7 +1105,7 @@ int32_t dp_update_subif_info(struct dp_subif_upd_info *info)
 			DP_DEBUG(DP_DBG_FLAG_REG,
 				 "reduce old bp refcnt:%d\n", bp_info->ref_cnt);
 			dp_notifier_invoke(info->inst,
-					   old_ctp_dev, info->dp_port,
+					   old_ctp_dev, info->dp_port, info->subif,
 					   DP_EVENT_DE_REGISTER_SUBIF);
 		}
 		old_data.ctp_dev = old_ctp_dev;
@@ -1117,7 +1117,7 @@ int32_t dp_update_subif_info(struct dp_subif_upd_info *info)
 			 old_dev->name);
 		dp_inst_del_dev(old_dev, info->inst, info->dp_port, vap, 0);
 
-		dp_notifier_invoke(info->inst, old_dev, info->dp_port,
+		dp_notifier_invoke(info->inst, old_dev, info->dp_port, info->subif,
 				   DP_EVENT_DE_REGISTER_SUBIF);
 		dp_free_bp_priv(info->inst, sif->bp);
 	}
@@ -1361,7 +1361,7 @@ int32_t dp_register_dev_ext(int inst, struct module *owner, u32 port_id,
 	DP_LIB_LOCK(&dp_lock);
 	if (flags & DP_F_DEREGISTER) {	/*de-register */
 		if (port_info->status == PORT_DEV_REGISTERED) {
-			dp_notifier_invoke(inst, port_info->dev, port_id,
+			dp_notifier_invoke(inst, port_info->dev, port_id, 0,
 					   DP_EVENT_DE_REGISTER_DEV);
 			port_info->status = PORT_ALLOCATED;
 			DP_CB(inst, dev_platform_set)(inst, port_id, data,
@@ -1438,7 +1438,7 @@ int32_t dp_register_dev_ext(int inst, struct module *owner, u32 port_id,
 	if (dp_cb)
 		port_info->cb = *dp_cb;
 
-	dp_notifier_invoke(inst, port_info->dev, port_id,
+	dp_notifier_invoke(inst, port_info->dev, port_id, 0,
 			   DP_EVENT_REGISTER_DEV);
 
 	DP_LIB_UNLOCK(&dp_lock);
diff --git a/drivers/net/datapath/dpm/datapath_misc.c b/drivers/net/datapath/dpm/datapath_misc.c
index 16bea00afb30..0b8761b6be30 100644
--- a/drivers/net/datapath/dpm/datapath_misc.c
+++ b/drivers/net/datapath/dpm/datapath_misc.c
@@ -1052,7 +1052,7 @@ int32_t dp_sync_subifid_priv(struct net_device *dev, char *subif_name,
 					subif_name, flags, subifid_fn);
 
 			if (get_dp_bp_info(0, subif_id->bport)->ref_cnt <= 1)
-				dp_notifier_invoke(0, dev, subif_id->port_id,
+				dp_notifier_invoke(0, dev, subif_id->port_id, subif_id->subif,
 						   DP_EVENT_REGISTER_SUBIF);
 
 			if (data && data->ctp_dev) {
@@ -1060,7 +1060,7 @@ int32_t dp_sync_subifid_priv(struct net_device *dev, char *subif_name,
 						&subif_id[1], NULL, flags,
 						subifid_fn);
 				dp_notifier_invoke(0, data->ctp_dev,
-						   subif_id->port_id,
+						   subif_id->port_id, subif_id->subif,
 						   DP_EVENT_REGISTER_SUBIF);
 			}
 		}
diff --git a/drivers/net/datapath/dpm/datapath_notifier.c b/drivers/net/datapath/dpm/datapath_notifier.c
index 91cab8cfecf9..3b524a300ba6 100644
--- a/drivers/net/datapath/dpm/datapath_notifier.c
+++ b/drivers/net/datapath/dpm/datapath_notifier.c
@@ -32,12 +32,10 @@ int unregister_notifier(u32 flag)
 }
 
 int dp_notifier_invoke(int inst, struct net_device *dev,
-		       u32 port_id, enum DP_EVENT_TYPE type)
+		       u32 port_id, u32 subif_id, enum DP_EVENT_TYPE type)
 {
 	struct pmac_port_info *port_info;
 	struct dp_evt_notif_data notif_data = {0};
-	dp_subif_t subif = {0};
-	int ret = DP_SUCCESS;
 
 	port_info = get_dp_port_info(inst, port_id);
 
@@ -58,14 +56,7 @@ int dp_notifier_invoke(int inst, struct net_device *dev,
 		break;
 	case DP_EVENT_REGISTER_SUBIF:
 	case DP_EVENT_DE_REGISTER_SUBIF:
-		ret = dp_get_netif_subifid_priv(inst, dev, NULL, NULL, NULL,
-						&subif, 0);
-		if (ret) {
-			pr_err("%s: failed to get subif info for %s\n",
-			       __func__, dev->name);
-			return DP_FAILURE;
-		}
-		notif_data.subif = subif.subif;
+		notif_data.subif = subif_id;
 		break;
 	default:
 		pr_err("%s Unsupported Event %d\n", __func__, type);
